### PR TITLE
8290781: Segfault at PhaseIdealLoop::clone_loop_handle_data_uses

### PR DESCRIPTION
**Pull request by rwestrel**
In the test case, both:

            v1 = array[i];
(in the first loop), and

            v += array[i-1];

(in the second one) access the same element. The bound check for the
second access is optimized out and the load of that access becomes
control dependent on the range check of the first one.

In the context of loop strip mining, data nodes that are in the outer
loop are expected to be reachable from the safepoint node. There are
rare case when it's not the case so I added logic to fix those cases
before loop cloning. That logic covers both nodes that have control in
the outer loop and control input in the outer loop. That logic is
incorrect (in the case of a node with a control input in the outer
loop): when cloning the first loop body, that logic finds the load
referenced from the range check in the loop body. That load has a
control input that's in the inner loop. But the logic only check
whether it's in the outer loop: anything in the inner loop is also in
the outer loop with that logic. The control of the load is then
wrongly updated to be outside the outer loop. That then causes a crash
because the load is recorded as being in the body of the second loop
but its control is not.

The fix I propose stengthen the logic that checks for nodes in the
outer loop: it checks that the control input is in the outer loop but
not in the inner loop.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290781](https://bugs.openjdk.org/browse/JDK-8290781): Segfault at PhaseIdealLoop::clone_loop_handle_data_uses

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [43dfe289](https://git.openjdk.org/jdk/pull/9997/files/43dfe289fa53208636df3fdab8781f0ded6cb02f)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [90938577](https://git.openjdk.org/jdk/pull/9997/files/90938577880b8648fb95012c9ee2ca51e7e93919)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9997/head:pull/9997` \
`$ git checkout pull/9997`

Update a local copy of the PR: \
`$ git checkout pull/9997` \
`$ git pull https://git.openjdk.org/jdk pull/9997/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9997`

View PR using the GUI difftool: \
`$ git pr show -t 9997`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9997.diff">https://git.openjdk.org/jdk/pull/9997.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/loopopts.cpp [hunk 1]
```diff
@@ -2160,11 +2160,18 @@ static void clone_outer_loop_helper(Node* n, const IdealLoopTree *loop, const Id
...
-          // the outer loop too
-          (u->in(0) != NULL && outer_loop->is_member(phase->get_loop(u->in(0))))) {
+      if (outer_loop->is_member(u_loop)) {
         wq.push(u);
+      } else {
+      // nodes pinned with control in the outer loop but not referenced from the safepoint must be moved out of
+      // the outer loop too
+        Node *u_c = u->in(0);
+        if (u_c != NULL) {
+          IdealLoopTree *u_c_loop = phase->get_loop(u_c);
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopopts.cpp:2170]:
```suggestion
          IdealLoopTree* u_c_loop = phase->get_loop(u_c);
```

**Review by chhagedorn — APPROVED** [review]
Looks good to me!

**Comment by rwestrel** [comment]
@chhagedorn thanks for the review and the formatting fixes

File: src/hotspot/share/opto/loopopts.cpp [hunk 2]
```diff
@@ -2160,11 +2160,18 @@ static void clone_outer_loop_helper(Node* n, const IdealLoopTree *loop, const Id
...
-      if (outer_loop->is_member(u_loop) ||
-          // nodes pinned with control in the outer loop but not referenced from the safepoint must be moved out of
-          // the outer loop too
-          (u->in(0) != NULL && outer_loop->is_member(phase->get_loop(u->in(0))))) {
+      if (outer_loop->is_member(u_loop)) {
         wq.push(u);
+      } else {
+      // nodes pinned with control in the outer loop but not referenced from the safepoint must be moved out of
+      // the outer loop too
+        Node* u_c = u->in(0);
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopopts.cpp:2168]:
```suggestion
        // nodes pinned with control in the outer loop but not referenced from the safepoint must be moved out of
        // the outer loop too
        Node* u_c = u->in(0);
```

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. I'll run some testing and report back once it passed.

**Comment by TobiHartmann** [comment]
All tests passed.

**Comment by rwestrel** [comment]
@TobiHartmann thanks for the review and testing

**Pull request closed without merging**